Add a module for sra-tools prefetch

Midnighter opened the pull request:
<!--
# nf-core/modules pull request

Many thanks for contributing to nf-core/modules!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the master branch.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
-->

Add a module for sra-tools prefetch which will be part of a sub-workflow that is later added to the fetchngs pipeline.

## PR checklist

Closes #XXX <!-- If this PR fixes an issue, please link it here! -->

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

drpatelh commented on modules/sratools/prefetch/main.nf:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    input:
+    val(meta)
```
```suggestion
    tuple val (meta), val (id)
```

drpatelh commented on modules/sratools/prefetch/main.nf:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    input:
+    val(meta)
```
```suggestion
    tuple val(meta), val(id)
```

drpatelh replied:
In theory this module should work just off an `id` that doesn't necessarily have to have any sample information stored in `map`. I wonder if the safer way of coding this is to have a separate `id` in the channel? This way developers can provide an empy `meta` if they don't have any sample information.

drpatelh commented on modules/sratools/prefetch/main.nf:
```diff
@@ -0,0 +1,38 @@
+    val(meta)
+
+    output:
+    tuple val(meta), path("${meta.id}/"), emit: reads
+    path "*.version.txt"          , emit: version
```
```suggestion
    tuple val(meta), path("$id}"), emit: reads
    path "*.version.txt"                , emit: version
```

drpatelh commented on modules/sratools/prefetch/main.nf:
```diff
@@ -0,0 +1,38 @@
+    prefetch \\
+        --progress \\
+        $options.args \\
+        ${meta.id}
```
```suggestion
        ${id}
```

drpatelh commented on modules/sratools/prefetch/main.nf:
```diff
@@ -0,0 +1,38 @@
+options        = initOptions(params.options)
+
+process SRATOOLS_PREFETCH {
+    tag "$meta.id"
```
```suggestion
    tag "$id"
```

drpatelh commented on modules/sratools/prefetch/main.nf:
```diff
@@ -0,0 +1,38 @@
+    label 'process_low'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
```
```suggestion
        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:[:], publish_by_meta:[]) }
```

Midnighter replied:
Adding this change causes the linting to fail.

```
'meta:meta' unspecified in saveAs function
'publish_by_meta:['id']' unspecified in saveAs function
```

drpatelh replied:
Ok. I thought it didn't make sense here to use `meta.id` (sample name) to publish the results because it should really be published under `id` if anything. But I also remember asking for that lint test to be added in :) 

Can you change it back please?

drpatelh commented on tests/modules/sratools/prefetch/main.nf:
```diff
@@ -0,0 +1,12 @@
+
+workflow test_sratools_prefetch {
+
+    input = [ id:'ERR2815334', single_end:false ] // meta map
```
```suggestion
    input = [ id:'ERR2815334', single_end:false ] // meta map
```
```suggestion
    input = [
        [ id:'my_sample', single_end:false ], // meta map
        'ERR2815334'
    ]
```

Midnighter commented:
I changed it to only requiring an `id` as you suggested.

Locally, the docker test runs perfectly fine. I'm not sure why the output is missing here.

What do you think about my open points in the opening comment regarding validation and retrying?

drpatelh commented on modules/sratools/prefetch/meta.yml:
```diff
@@ -0,0 +1,39 @@
+      description: >
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - id:
+      type: directory
+      description: >
+        Directory containing the ETL data for the given SRA id.
+      pattern: "meta.id"
```
This is an `input` and not an `output`

drpatelh replied:
And need an entry for `sra` here too in `output`

drpatelh commented on modules/sratools/prefetch/main.nf:
```diff
@@ -0,0 +1,38 @@
+    tuple val(meta), val(id)
+
+    output:
+    tuple val(meta), path("$id"), emit: sra
+    path "*.version.txt"          , emit: version
```
```suggestion
    tuple val(meta), path("$id"), emit: sra
    path "*.version.txt"        , emit: version
```

JoseEspinosa commented:
Hi @Midnighter 
I find this pipeline having the same issue, see [here](https://github.com/TheJacksonLaboratory/splicing-pipelines-nf/pull/253#issuecomment-906543090).
Seems that the solution is to provide a configuration file, see [here](https://github.com/TheJacksonLaboratory/splicing-pipelines-nf/pull/253/commits/2fce4f25b2af90f2db20921baed7935e3a7f03bc)
Hope it helps!

Midnighter commented:
Addressed your comments. So just need to figure out something for retrying and decide on validating the downloaded data. If I understand the help text correctly, some simple verification is done by prefetch already.

```
  -C|--verify <yes|no>             Verify after download - one of: no, yes
                                     [default]
```
